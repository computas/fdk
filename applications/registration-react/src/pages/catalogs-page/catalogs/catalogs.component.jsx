import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import { CatalogItem } from './catalog-item/catalog-item.component';

export const Catalog = props => {
  const { catalogId, type, items, fetchItems } = props;
  fetchItems(catalogId);
  return (
    <CatalogItem
      key={catalogId}
      publisherId={catalogId}
      type={type}
      itemsCount={_.get(items, [catalogId, 'items'], []).length}
    />
  );
};

Catalog.defaultProps = {
  catalogId: null,
  items: false,
  fetchItems: _.noop
};

Catalog.propTypes = {
  catalogId: PropTypes.string,
  type: PropTypes.string.isRequired,
  items: PropTypes.object,
  fetchItems: PropTypes.func
};
