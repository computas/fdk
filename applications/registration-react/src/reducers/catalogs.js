import {
  CATALOGS_REQUEST,
  CATALOGS_SUCCESS,
  CATALOGS_FAILURE
} from '../constants/ActionTypes';

export default function catalogs(
  state = { isFetchingCatalogs: false, catalogItems: null },
  action
) {
  switch (action.type) {
    case CATALOGS_REQUEST:
      return {
        ...state,
        isFetchingCatalogs: true
      };
    case CATALOGS_SUCCESS:
      return {
        ...state,
        isFetchingCatalogs: false,
        catalogItems: action.response.data
      };
    case CATALOGS_FAILURE:
      return {
        ...state,
        isFetchingCatalogs: false,
        catalogItems: null
      };
    default:
      return state;
  }
}
