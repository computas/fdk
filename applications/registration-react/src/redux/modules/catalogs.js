import _ from 'lodash';
import {
  CATALOGS_REQUEST,
  CATALOGS_SUCCESS,
  CATALOGS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchCatalogsIfNeeded(catalogsURL) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), 'catalogs')) &&
    dispatch(
      fetchActions(catalogsURL, [
        CATALOGS_REQUEST,
        CATALOGS_SUCCESS,
        CATALOGS_FAILURE
      ])
    );
}

const initialState = { isFetchingCatalogs: false, catalogItems: null };

export default function catalogs(state = initialState, action) {
  switch (action.type) {
    case CATALOGS_REQUEST:
      return {
        ...state,
        isFetching: true,
        lastFetch: null
      };
    case CATALOGS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        lastFetch: Date.now(),
        catalogItems: _.get(action.payload, ['_embedded', 'catalogs'])
      };
    case CATALOGS_FAILURE:
      return {
        ...state,
        isFetching: false,
        lastFetch: null,
        catalogItems: null
      };
    default:
      return state;
  }
}

export const getCatalogItemByCatalogId = (catalogItems, catalogId) => {
  if (Array.isArray(_.get(catalogItems, ['_embedded', 'catalogs']))) {
    return _.get(catalogItems, ['_embedded', 'catalogs']).filter(
      item => item.id === catalogId
    )[0];
  }
  return null;
};
