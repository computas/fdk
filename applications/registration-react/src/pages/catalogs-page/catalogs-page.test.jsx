import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from 'redux-test-utils';
import shallowWithStore from '../../../test/shallowWithStore';
import RegCatalogsComponent, { RegCatalogs } from './catalogs-page';
import catalogs from '../../../test/fixtures/catalogs';

let defaultProps;
let wrapper;
let dispatch;

beforeEach(() => {
  dispatch = jest.fn();
  defaultProps = {
    dispatch,
    catalogItems: catalogs.catalogItems,
    isFetchingCatalogs: false
  };
  wrapper = shallow(<RegCatalogs {...defaultProps} />);
});

test('should render RegCatalogsComponent correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render RegCatalogsComponent correctly with catalogItems', () => {
  const testState = {
    catalogs
  };
  const store = createMockStore(testState);
  wrapper = shallowWithStore(<RegCatalogsComponent />, store);
  expect(wrapper).toMatchSnapshot();
});
