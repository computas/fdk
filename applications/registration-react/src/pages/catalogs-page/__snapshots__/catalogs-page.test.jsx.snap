// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RegCatalogsComponent correctly 1`] = `
<div
  className="container"
>
  <div
    className="row mb-2 mb-md-5"
    key="910244132"
  >
    <div
      className="col-12"
    >
      <div>
        <h2
          className="fdk-text-strong mb-4"
        >
          RAMSUND OG ROGNAN REVISJON
        </h2>
      </div>
      <CardGroup
        tag="div"
      >
        <FeatureToggle
          featureName="API"
          showOnlyWhenDisabled={false}
        />
      </CardGroup>
    </div>
  </div>
</div>
`;

exports[`should render RegCatalogsComponent correctly with catalogItems 1`] = `
<RegCatalogs
  catalogItems={
    Array [
      Object {
        "description": Object {},
        "id": "910244132",
        "publisher": Object {
          "name": "RAMSUND OG ROGNAN REVISJON",
          "uri": "http://data.brreg.no/enhetsregisteret/enhet/910244132",
        },
        "title": Object {
          "nb": "Datakatalog for RAMSUND OG ROGNAN REVISJON",
        },
        "uri": "http://brreg.no/catalogs/910244132",
      },
    ]
  }
  fetchCatalogsIfNeeded={[Function]}
  fetchDatasetsIfNeeded={[Function]}
  isFetching={false}
/>
`;
